src/rulesets/Base/array.jl

```diff
@@ -196,7 +199,9 @@ function rrule(::typeof(hvcat), rows, values::Union{AbstractArray, Number}...)
             end
             dY[index...]
         end
-        return (NoTangent(), NoTangent(), dXs...)
+        _call((f, x)) = f(x)
+        proj_dXs = _call.(tuple(zip(project_Vs, dXs)...)) # can't make it type infer, e.g. test_rrule(hvcat, 1, rand(3)', transpose(rand(3)) ⊢ rand(1,3))
```

Comment:
This is pretty ugly.
Isn;t it the same as 
```suggestion
        _call(f, x) = f(x)
        proj_dXs = map(_call, project_Vs, dXs)
```
Which may or may not fix inference, but is at least more readable.

Reply:
It's like that because a vector of functions does not infer, but a tuple of functions does. e.g. see
```julia
julia> @code_warntype map(zip([-, +], [1.0, 2.0])) do (f, b)
           return f(b)
       end
```
vs
```julia
julia> @code_warntype map(zip((-, +), [1.0, 2.0])) do (f, b)
           return f(b)
       end
```
(but that alone dos not seem to be enough)

Reply:
right but `map` over two tuples does infer.
```julia
julia> @code_warntype map((-, +), (1.0, 2.0)) do f, b
                  return f(b)
              end
MethodInstance for map(::var"#41#42", ::Tuple{typeof(-), typeof(+)}, ::Tuple{Float64, Float64})
  from map(f, t::Tuple{Any, Any}, s::Tuple{Any, Any}) in Base at tuple.jl:247
Arguments
  #self#::Core.Const(map)
  f::Core.Const(var"#41#42"())
  t::Core.Const((-, +))
  s::Tuple{Float64, Float64}
Body::Tuple{Float64, Float64}
1 ─      nothing
│   %2 = Base.getindex(t, 1)::Core.Const(-)
│   %3 = Base.getindex(s, 1)::Float64
│   %4 = (f)(%2, %3)::Float64
│   %5 = Base.getindex(t, 2)::Core.Const(+)
│   %6 = Base.getindex(s, 2)::Float64
│   %7 = (f)(%5, %6)::Float64
│   %8 = Core.tuple(%4, %7)::Tuple{Float64, Float64}
└──      return %8
```

Reply:
From what I can tell this fixes it.
I will push these changes